File: 0-mock-server/master-ws-server-dual/src/DeviceConnectionManager.ts
import WebSocket from 'ws';
import {DeviceType, DeviceInfo, DeviceRegistration, RuntimeConfig} from './types';
import {DEFAULT_RUNTIME_CONFIG, mergeRuntimeConfig} from './config';

interface ConnectedDevice {
  socket: WebSocket;
  info: DeviceInfo;
  lastHeartbeat: number;
}

/** 一对一配对：master deviceId -> { master, slave, runtimeConfig } */
interface DevicePair {
  master?: ConnectedDevice;
  slave?: ConnectedDevice;
  runtimeConfig: RuntimeConfig;
}

export class DeviceConnectionManager {
  /** token -> DeviceInfo */
  private pendingDevices = new Map<string, DeviceInfo>();
  /** masterDeviceId -> DevicePair */
  private pairs = new Map<string, DevicePair>();
  /** socket -> deviceId（快速反查） */
  private socketToDeviceId = new Map<WebSocket, string>();
  /** slave deviceId -> masterDeviceId（快速反查） */
  private slaveToMaster = new Map<string, string>();

  preRegisterDevice(registration: DeviceRegistration, token: string): { success: boolean; error?: string } {
    const {type, deviceId, masterDeviceId} = registration;

    if (type === DeviceType.MASTER) {
      // master 不能重复注册
      const pair = this.pairs.get(deviceId);
      if (pair?.master) {
        return {success: false, error: 'Master device already connected'};
      }
    } else {
      if (!masterDeviceId) {
        return {success: false, error: 'masterDeviceId is required for slave device'};
      }
      const pair = this.pairs.get(masterDeviceId);
      if (!pair?.master) {
        return {success: false, error: 'Master device not connected'};
      }
      if (pair.slave) {
        return {success: false, error: 'This master already has a slave connected'};
      }
    }

    this.pendingDevices.set(token, {
      ...registration,
      connectedAt: new Date(),
      token
    });

    return {success: true};
  }

  connectDeviceWithToken(socket: WebSocket, token: string): { success: boolean; error?: string; deviceInfo?: DeviceInfo } {
    const deviceInfo = this.pendingDevices.get(token);
    if (!deviceInfo) {
      return {success: false, error: 'Invalid or expired token'};
    }

    const {type, deviceId, masterDeviceId} = deviceInfo;
    this.pendingDevices.delete(token);
    deviceInfo.connectedAt = new Date();

    const connected: ConnectedDevice = {socket, info: deviceInfo, lastHeartbeat: Date.now()};

    if (type === DeviceType.MASTER) {
      const existing = this.pairs.get(deviceId);
      if (existing?.master) {
        return {success: false, error: 'Master device already connected'};
      }
      const pair: DevicePair = existing ?? {runtimeConfig: mergeRuntimeConfig(deviceInfo.runtimeConfig)};
      pair.master = connected;
      this.pairs.set(deviceId, pair);
    } else {
      const pair = this.pairs.get(masterDeviceId!);
      if (!pair?.master) {
        return {success: false, error: 'Master device disconnected'};
      }
      if (pair.slave) {
        return {success: false, error: 'This master already has a slave'};
      }
      pair.slave = connected;
      this.slaveToMaster.set(deviceId, masterDeviceId!);
    }

    this.socketToDeviceId.set(socket, deviceId);
    return {success: true, deviceInfo};
  }

  getRuntimeConfig(masterDeviceId: string): RuntimeConfig {
    return this.pairs.get(masterDeviceId)?.runtimeConfig ?? DEFAULT_RUNTIME_CONFIG;
  }

  findDeviceBySocket(socket: WebSocket): { type: DeviceType; deviceId: string; masterDeviceId: string } | undefined {
    const deviceId = this.socketToDeviceId.get(socket);
    if (!deviceId) return undefined;

    // master: deviceId 就是 masterDeviceId
    const pair = this.pairs.get(deviceId);
    if (pair?.master?.socket === socket) {
      return {type: DeviceType.MASTER, deviceId, masterDeviceId: deviceId};
    }

    // slave: 通过 slaveToMaster 直接查
    const masterId = this.slaveToMaster.get(deviceId);
    if (masterId) {
      return {type: DeviceType.SLAVE, deviceId, masterDeviceId: masterId};
    }

    return undefined;
  }

  getPeer(masterDeviceId: string, selfType: DeviceType): ConnectedDevice | undefined {
    const pair = this.pairs.get(masterDeviceId);
    if (!pair) return undefined;
    return selfType === DeviceType.MASTER ? pair.slave : pair.master;
  }

  getSlave(masterDeviceId: string): ConnectedDevice | undefined {
    return this.pairs.get(masterDeviceId)?.slave;
  }

  getMaster(masterDeviceId: string): ConnectedDevice | undefined {
    return this.pairs.get(masterDeviceId)?.master;
  }

  disconnectSlave(masterDeviceId: string) {
    const pair = this.pairs.get(masterDeviceId);
    if (!pair?.slave) return;
    this.socketToDeviceId.delete(pair.slave.socket);
    this.slaveToMaster.delete(pair.slave.info.deviceId);
    if (pair.slave.socket.readyState === WebSocket.OPEN) {
      pair.slave.socket.close();
    }
    pair.slave = undefined;
  }

  disconnectMaster(masterDeviceId: string) {
    const pair = this.pairs.get(masterDeviceId);
    if (!pair) return;

    // 先断 slave
    if (pair.slave) {
      this.socketToDeviceId.delete(pair.slave.socket);
      this.slaveToMaster.delete(pair.slave.info.deviceId);
      if (pair.slave.socket.readyState === WebSocket.OPEN) {
        pair.slave.socket.close();
      }
    }
    // 再断 master
    if (pair.master) {
      this.socketToDeviceId.delete(pair.master.socket);
    }
    this.pairs.delete(masterDeviceId);
  }

  updateHeartbeat(socket: WebSocket) {
    const device = this.findDeviceBySocket(socket);
    if (!device) return;
    const pair = this.pairs.get(device.masterDeviceId);
    if (!pair) return;
    const target = device.type === DeviceType.MASTER ? pair.master : pair.slave;
    if (target) target.lastHeartbeat = Date.now();
  }

  checkAndDisconnectTimeoutDevices(): Array<{ type: DeviceType; deviceId: string; masterDeviceId: string }> {
    const result: Array<{ type: DeviceType; deviceId: string; masterDeviceId: string }> = [];
    const now = Date.now();

    for (const [masterId, pair] of this.pairs) {
      const timeout = pair.runtimeConfig.heartbeatTimeout;
      if (pair.master && now - pair.master.lastHeartbeat > timeout) {
        result.push({type: DeviceType.MASTER, deviceId: pair.master.info.deviceId, masterDeviceId: masterId});
        pair.master.socket.close();
      }
      if (pair.slave && now - pair.slave.lastHeartbeat > timeout) {
        result.push({type: DeviceType.SLAVE, deviceId: pair.slave.info.deviceId, masterDeviceId: masterId});
        pair.slave.socket.close();
      }
    }
    return result;
  }

  cleanExpiredPendingDevices(expireTime: number) {
    const now = Date.now();
    for (const [token, info] of this.pendingDevices) {
      if (now - info.connectedAt.getTime() > expireTime) {
        this.pendingDevices.delete(token);
      }
    }
  }

  getAllSockets(): Array<{ socket: WebSocket; masterDeviceId: string }> {
    const result: Array<{ socket: WebSocket; masterDeviceId: string }> = [];
    for (const [masterId, pair] of this.pairs) {
      if (pair.master) result.push({socket: pair.master.socket, masterDeviceId: masterId});
      if (pair.slave) result.push({socket: pair.slave.socket, masterDeviceId: masterId});
    }
    return result;
  }

  getStats() {
    let masterCount = 0, slaveCount = 0;
    const pairList: Array<{ masterDeviceId: string; slaveDeviceId?: string }> = [];

    for (const [masterId, pair] of this.pairs) {
      if (pair.master) masterCount++;
      if (pair.slave) slaveCount++;
      pairList.push({
        masterDeviceId: masterId,
        slaveDeviceId: pair.slave?.info.deviceId
      });
    }

    return {
      masterCount,
      slaveCount,
      pendingCount: this.pendingDevices.size,
      pairs: pairList
    };
  }
}

File: 3-adapter/android/adapterPure/adapter-lib/src/main/java/com/impos2/adapter/webserver/DeviceConnectionManager.kt
package com.impos2.adapter.webserver

import android.util.Log
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

data class RuntimeConfig(
    val heartbeatTimeout: Long = 60_000L,
    val retryCacheTimeout: Long = 30_000L,
)

data class ServerConfig(
    val port: Int = 8888,
    val basePath: String = "/localServer",
    val heartbeatInterval: Long = 30_000L,
    val defaultRuntimeConfig: RuntimeConfig = RuntimeConfig(),
)

enum class DeviceType { MASTER, SLAVE }

data class DeviceInfo(
    val type: DeviceType,
    val deviceId: String,
    val masterDeviceId: String,
    val token: String,
    val runtimeConfig: RuntimeConfig? = null,
    val connectedAt: Long = System.currentTimeMillis(),
)

data class DeviceConnectionStats(
    val masterCount: Int,
    val slaveCount: Int,
    val pendingCount: Int,
    val uptime: Long,
)

/**
 * 设备连接关系管理器。
 *
 * 这个类只关注“设备之间如何建立、维持、清理连接关系”，不关心 WebSocket/HTTP 细节。
 * LocalWebServer 收到的注册、心跳、配对、断开事件最终都会落到这里，统一形成：
 * - pending 设备；
 * - 已配对 master/slave 关系；
 * - socket 与设备 id 的映射；
 * - 可观测的连接统计。
 *
 * 设计重点：
 * - 写路径尽量短，避免在锁里做 I/O；
 * - 读路径保持快照一致性，便于诊断与状态查询；
 * - 对异常顺序的事件具备容错能力，避免脏映射越积越多。
 */
class DeviceConnectionManager(private val config: ServerConfig) {

    companion object {
        private const val TAG = "DeviceConnectionManager"
        private const val TOKEN_EXPIRE_MS = 5 * 60 * 1000L
    }

    // 保护多张映射表的读写一致性。锁内只做轻量状态变更，不做网络 I/O。
    private val lock = ReentrantReadWriteLock()
    private val pending = ConcurrentHashMap<String, DeviceInfo>()
    private val pairs = ConcurrentHashMap<String, DevicePair>()
    private val socketToDevice = ConcurrentHashMap<String, String>()
    private val slaveToMaster = ConcurrentHashMap<String, String>()
    private val startTime = System.currentTimeMillis()
    private val connectCount = AtomicLong(0L)
    private val disconnectCount = AtomicLong(0L)
    private val heartbeatUpdateCount = AtomicLong(0L)
    private val pendingCleanupCount = AtomicLong(0L)
    private val rejectedPreRegisterCount = AtomicLong(0L)
    private val rejectedConnectCount = AtomicLong(0L)

    data class ConnectedDevice(
        val socket: WsSession,
        val info: DeviceInfo,
        @Volatile var lastHeartbeat: Long = System.currentTimeMillis(),
    )

    data class DevicePair(
        @Volatile var master: ConnectedDevice? = null,
        @Volatile var slave: ConnectedDevice? = null,
        val runtimeConfig: RuntimeConfig,
    )

    fun preRegister(info: DeviceInfo): String? = lock.write {
        val validationError = validatePreRegisterLocked(info)
        if (validationError != null) {
            rejectedPreRegisterCount.incrementAndGet()
            Log.w(TAG, "preRegister rejected: type=${info.type}, deviceId=${info.deviceId}, reason=$validationError")
            return validationError
        }

        removeDuplicatePendingLocked(info)
        pending[info.token] = info
        null
    }

    fun connect(socket: WsSession, token: String): Pair<DeviceInfo?, String?> = lock.write {
        val info = pending.remove(token) ?: run {
            rejectedConnectCount.incrementAndGet()
            return null to "Invalid or expired token"
        }
        val connected = ConnectedDevice(socket, info)

        if (info.type == DeviceType.MASTER) {
            val existing = pairs[info.deviceId]
            if (existing?.master != null) {
                rejectedConnectCount.incrementAndGet()
                return null to "Master already connected"
            }
            val runtimeConfig = info.runtimeConfig ?: config.defaultRuntimeConfig
            val pair = existing ?: DevicePair(runtimeConfig = runtimeConfig)
            pair.master = connected
            pairs[info.deviceId] = pair
        } else {
            val pair = pairs[info.masterDeviceId] ?: run {
                rejectedConnectCount.incrementAndGet()
                return null to "Master disconnected"
            }
            if (pair.master == null) {
                rejectedConnectCount.incrementAndGet()
                return null to "Master disconnected"
            }
            if (pair.slave != null) {
                rejectedConnectCount.incrementAndGet()
                return null to "Master already has a slave"
            }
            pair.slave = connected
            slaveToMaster[info.deviceId] = info.masterDeviceId
        }

        socketToDevice[socket.key] = info.deviceId
        connectCount.incrementAndGet()
        Log.i(TAG, "connect: type=${info.type}, deviceId=${info.deviceId}, master=${info.masterDeviceId}")
        info to null
    }

    fun getRuntimeConfig(masterDeviceId: String): RuntimeConfig = lock.read {
        pairs[masterDeviceId]?.runtimeConfig ?: config.defaultRuntimeConfig
    }

    fun findBySocket(socketKey: String): Triple<DeviceType, String, String>? = lock.read {
        val deviceId = socketToDevice[socketKey] ?: return null
        val pair = pairs[deviceId]
        if (pair?.master?.socket?.key == socketKey) {
            return Triple(DeviceType.MASTER, deviceId, deviceId)
        }
        val masterId = slaveToMaster[deviceId] ?: return null
        Triple(DeviceType.SLAVE, deviceId, masterId)
    }

    fun getPeer(masterDeviceId: String, selfType: DeviceType): ConnectedDevice? = lock.read {
        val pair = pairs[masterDeviceId] ?: return null
        if (selfType == DeviceType.MASTER) pair.slave else pair.master
    }

    fun getMaster(masterDeviceId: String): ConnectedDevice? = lock.read {
        pairs[masterDeviceId]?.master
    }

    fun getSlave(masterDeviceId: String): ConnectedDevice? = lock.read {
        pairs[masterDeviceId]?.slave
    }

    fun updateHeartbeat(socketKey: String) {
        lock.read {
            val deviceId = socketToDevice[socketKey] ?: return
            val masterPair = pairs[deviceId]
            if (masterPair?.master?.socket?.key == socketKey) {
                masterPair.master?.lastHeartbeat = System.currentTimeMillis()
                heartbeatUpdateCount.incrementAndGet()
                return
            }
            val masterId = slaveToMaster[deviceId] ?: return
            val pair = pairs[masterId] ?: return
            pair.slave?.takeIf { it.socket.key == socketKey }?.let {
                it.lastHeartbeat = System.currentTimeMillis()
                heartbeatUpdateCount.incrementAndGet()
            }
        }
    }

    fun disconnectMaster(masterDeviceId: String) {
        val sessionsToClose = lock.write {
            val pair = pairs.remove(masterDeviceId) ?: return
            val sessions = mutableListOf<WsSession>()
            pair.slave?.let {
                removeSocketAssociationsLocked(it)
                sessions += it.socket
            }
            pair.master?.let {
                removeSocketAssociationsLocked(it)
                sessions += it.socket
            }
            disconnectCount.incrementAndGet()
            Log.i(TAG, "disconnectMaster: master=$masterDeviceId")
            sessions
        }
        closeSessions(sessionsToClose)
    }

    fun disconnectSlave(masterDeviceId: String) {
        val sessionToClose = lock.write {
            val pair = pairs[masterDeviceId] ?: return
            val slave = pair.slave ?: return
            removeSocketAssociationsLocked(slave)
            pair.slave = null
            disconnectCount.incrementAndGet()
            Log.i(TAG, "disconnectSlave: master=$masterDeviceId, slave=${slave.info.deviceId}")
            slave.socket
        }
        closeSession(sessionToClose)
    }

    fun checkHeartbeatTimeouts(): List<Triple<DeviceType, String, String>> = lock.read {
        val now = System.currentTimeMillis()
        val result = mutableListOf<Triple<DeviceType, String, String>>()
        for ((masterId, pair) in pairs) {
            val timeout = pair.runtimeConfig.heartbeatTimeout
            pair.master?.let {
                if (now - it.lastHeartbeat > timeout) {
                    result.add(Triple(DeviceType.MASTER, it.info.deviceId, masterId))
                }
            }
            pair.slave?.let {
                if (now - it.lastHeartbeat > timeout) {
                    result.add(Triple(DeviceType.SLAVE, it.info.deviceId, masterId))
                }
            }
        }
        result
    }

    fun cleanExpiredPending() {
        val now = System.currentTimeMillis()
        val removed = lock.write {
            val expiredTokens = pending.entries
                .filter { now - it.value.connectedAt > TOKEN_EXPIRE_MS }
                .map { it.key }
            expiredTokens.forEach { pending.remove(it) }
            expiredTokens.size
        }
        if (removed > 0) {
            pendingCleanupCount.addAndGet(removed.toLong())
            Log.i(TAG, "cleanExpiredPending: removed=$removed")
        }
    }

    fun getAllSessions(): List<Pair<WsSession, String>> = lock.read {
        val result = mutableListOf<Pair<WsSession, String>>()
        for ((masterId, pair) in pairs) {
            pair.master?.let { result.add(it.socket to masterId) }
            pair.slave?.let { result.add(it.socket to masterId) }
        }
        result
    }

    fun getStats(): DeviceConnectionStats = lock.read {
        DeviceConnectionStats(
            masterCount = pairs.values.count { it.master != null },
            slaveCount = pairs.values.count { it.slave != null },
            pendingCount = pending.size,
            uptime = System.currentTimeMillis() - startTime,
        )
    }

    fun dumpState(): String = lock.read {
        buildString {
            append("masters=")
            append(pairs.values.count { it.master != null })
            append(", slaves=")
            append(pairs.values.count { it.slave != null })
            append(", pending=")
            append(pending.size)
            append(", sockets=")
            append(socketToDevice.size)
            append(", slaveMap=")
            append(slaveToMaster.size)
            append(", connects=")
            append(connectCount.get())
            append(", disconnects=")
            append(disconnectCount.get())
            append(", heartbeatUpdates=")
            append(heartbeatUpdateCount.get())
            append(", pendingCleanups=")
            append(pendingCleanupCount.get())
            append(", rejectedPreRegisters=")
            append(rejectedPreRegisterCount.get())
            append(", rejectedConnects=")
            append(rejectedConnectCount.get())
        }
    }

    fun close() {
        val sessions = lock.write {
            val result = mutableListOf<WsSession>()
            pairs.values.forEach { pair ->
                pair.master?.let {
                    removeSocketAssociationsLocked(it)
                    result += it.socket
                }
                pair.slave?.let {
                    removeSocketAssociationsLocked(it)
                    result += it.socket
                }
            }
            pairs.clear()
            pending.clear()
            socketToDevice.clear()
            slaveToMaster.clear()
            result
        }
        closeSessions(sessions)
    }

    private fun validatePreRegisterLocked(info: DeviceInfo): String? {
        return if (info.type == DeviceType.MASTER) {
            if (pairs[info.deviceId]?.master != null) {
                "Master already connected"
            } else {
                null
            }
        } else {
            val pair = pairs[info.masterDeviceId] ?: return "Master not connected"
            if (pair.master == null) {
                "Master not connected"
            } else if (pair.slave != null) {
                "Master already has a slave"
            } else {
                null
            }
        }
    }

    private fun removeDuplicatePendingLocked(info: DeviceInfo) {
        pending.entries.removeIf { entry ->
            val value = entry.value
            value.type == info.type &&
                value.deviceId == info.deviceId &&
                value.masterDeviceId == info.masterDeviceId &&
                entry.key != info.token
        }
    }

    private fun removeSocketAssociationsLocked(connected: ConnectedDevice) {
        socketToDevice.remove(connected.socket.key)
        if (connected.info.type == DeviceType.SLAVE) {
            slaveToMaster.remove(connected.info.deviceId)
        }
    }

    private fun closeSessions(sessions: List<WsSession>) {
        sessions.forEach { closeSession(it) }
    }

    private fun closeSession(session: WsSession?) {
        try {
            session?.close()
        } catch (e: Exception) {
            Log.w(TAG, "closeSession failed", e)
        }
    }
}
